Make loading weights 10-100x faster

jart opened the pull request:
This is a breaking change that's going to give us three benefits:

1. Your inference commands should load 100x faster
2. You may be able to safely load models 2x larger
3. You can run many concurrent inference processes

This was accomplished by changing the file format so we can mmap()
weights directly into memory without having to read() or copy them
thereby ensuring the kernel can make its file cache pages directly
accessible to our inference processes; and secondly, that the file
cache pages are much less likely to get evicted (which would force
loads to hit disk) because they're no longer competing with memory
pages that were needlessly created by gigabytes of standard i/o.

The new file format supports single-file models like LLaMA 7b, and
it also supports multi-file models like LLaMA 13B. Our Python tool
now merges the foo.1, foo.2, etc. files back into a single file so
that the C++ code which maps it doesn't need to reshape data every
time. That's made llama.cpp so much simpler. Much of its load code
has now been deleted.

Furthermore, this change ensures that tensors are aligned properly
on a 32-byte boundary. That opens the door to seeing if we can get
additional performance gains on some microprocessors, by using ops
that require memory alignment.

Lastly note that both POSIX and the Windows platform are supported

The issue this PR solves is https://github.com/ggerganov/llama.cpp/issues/91

This PR was written in collaboration with @slaren. This PR is also rebased on
PR #586 **so please do not squash merge**! Use either merge or rebase.

slaren committed:
Add mmap support for model files

slaren committed:
Fix ggml_init_params in quantize

slaren committed:
Make mmap_file static

slaren committed:
Unmap the file in llama_free

slaren committed:
Always initialize mm_addr and mm_length in llama_model

slaren committed:
Initial windows support (untested)

jart committed:
Make loading weights 10-100x faster

This is a breaking change that's going to give you three benefits:

1. Your inference commands should load 100x faster
2. You may be able to safely load models 2x larger
3. You can run many concurrent inference processes

This was accomplished by changing the file format so we can mmap()
weights directly into memory without having to read() or copy them
thereby ensuring the kernel can make its file cache pages directly
accessible to our inference processes; and secondly, that the file
cache pages are much less likely to get evicted (which would force
loads to hit disk) because they're no longer competing with memory
pages that were needlessly created by gigabytes of standard i/o.

The new file format supports single-file models like LLaMA 7b, and
it also supports multi-file models like LLaMA 13B. Our Python tool
now merges the foo.1, foo.2, etc. files back into a single file so
that the C++ code which maps it doesn't need to reshape data every
time. That's made llama.cpp so much simpler. Much of its load code
has now been deleted.

Furthermore, this change ensures that tensors are aligned properly
on a 32-byte boundary. That opens the door to seeing if we can get
additional performance gains on some microprocessors, by using ops
that require memory alignment.

Lastly note that both POSIX and the Windows platform are supported

Fixes #91

jart commented:
Yes indeed. I just fixed the quantize program. Now I'm hunting down all the tests.

jart commented:
All tests look green except for a CMake test. For example: https://github.com/ggerganov/llama.cpp/actions/runs/4559537462/jobs/8043597142?pr=613 I'm stumped on this error. I can't figure out where the file `models/ggml-vocab.bin` comes from. Does anyone know? Could it be a stale cache?

FNsi commented:
> All tests look green except for a CMake test. For example: https://github.com/ggerganov/llama.cpp/actions/runs/4559537462/jobs/8043597142?pr=613 I'm stumped on this error. I can't figure out where the file `models/ggml-vocab.bin` comes from. Does anyone know? Could it be a stale cache?

#355

bakkot commented on llama.h:
```diff
@@ -20,7 +20,7 @@
 #endif
 
 #define LLAMA_FILE_VERSION 1
-#define LLAMA_FILE_MAGIC 0x67676d66 // 'ggmf' in hex
```
Nit: why change the magic rather than the version? I assumed the plan was to keep the magic constant forever. If you bump the version instead, old executables will recognize new model files and [give a more useful error message](https://github.com/ggerganov/llama.cpp/blob/9cbc404ba6699a9ba4925ea25a60552b13491c7a/llama.cpp#L336). And it's nice to distinguish between "this is definitely model file for this project, but it's the wrong version" vs "this is some random junk we don't know anything about".

(This PR is a very neat bit of engineering; please don't let my nitpick distract from that.)

Green-Sky replied:
not a nitpick but a real change request :)

ggerganov commented:
@jart 
The `models/ggml-vocab.bin` is generated by `convert-pth-to-ggml.py` by providing an extra arg.

I had the expectation that `mmap` support would be much more intrusive, but in fact it turned out to be very compact. `llama.cpp` is much more simpler now. Good stuff

Regarding the version comment - yes, the plan was to bump versions and no the magic. But I'm ok to change the magic to commemorate the significance of this update. In fact, maybe we can make this a thing and everybody who makes a significant contribution to the project will get their initial appended to the version. What do you think? 😄 

Let me play with this tonight before merging. We have to make special care that all the other `ggml` model files floating around (Alpaca, GPT4All, Chinese LLaMA, etc.) have a nice way to convert to this new format and update the instructions in the README.

Also, maybe some synchronisation with #545 would be needed

jart commented:
File updated. A lot more tests are green now. No idea what's up with the sanitizer.

I thought so too! I too was pleasantly surprised by how well it worked out. Glad we took a few weeks to think.

I'm honored to hear you say that. I can roundup the magic to 64 bytes if you like, so there's room to hand out kudos without breaking backwards compatibility in the future. Since my initials also act as a stamp of approval, I'm going to be sending a follow-up change after this, that'll harden the loading code, so that folks will be able to trade model files for this format on HuggingFace with maximum safety and confidence.

#545 is an ambitious unification. I've done my best to comment my changes to make the merge less painful for the author. I've sought to update the other scripts too, but don't know how to run them. One thing you could also consider with this project is having a `contrib/` folder, where folks can merge as much of their own stuff as they want, under the expectation that the ones who need it are the ones who maintain it.

jart committed:
Ensure --mlock works properly with mmap() support

mqy commented on llama.cpp:
```diff
@@ -288,6 +307,57 @@ struct llama_context_params llama_context_default_params() {
+    int fd = open(fname, O_RDONLY);
+    if (fd == -1) return 0;
+    int64_t length = lseek(fd, 0, SEEK_END);
+    void *addr = mmap(NULL, length, PROT_READ, MAP_SHARED, fd, 0);
```
1. Is it more safe to use `mmap64` for 4GB+ files?
2. It seems `mmap`, `mmap64` and `MapViewOfFile` support mapping from given offset. Is it possible to map from header_len (as `offset`)? If we can do this, no need to align model file, right?

jart replied:
1. The right thing to do on 32-bit platforms is to have your build system define `-D_FILE_OFFSET_BITS=64` which will cause your system header files to automatically `#define mmap mmap64`
2. File offsets passed to mmap() need to be page size aligned, so I don't think so.

pgoodman replied:
@jart Is it possible to ensure the file size is a multiple of the hugepage size (e.g. using ftruncate), to benefit from fewer TLB lookups when the model data is accessed?

jart replied:
It doesn't matter with mmap() if the file length isn't page size aligned, even with smaller pages. You should be good to go if you modify the mmap() code in llama.cpp by hand and actually manage to get huge pages to work without nuking your machine :-)

pgoodman replied:
TIL!

jart committed:
Introduce GGML migration tool for new file format

If you deleted your old Meta LLaMA .pth files, then the
migrate-ggml-2023-03-30-pr613.py script will allow you to convert your
old ggml files into the new mmap()'able format.

See #613

jart commented on convert-pth-to-ggml.py:
```diff
@@ -24,16 +24,64 @@
+
+GGML_TYPE_SIZE = {
+    GGML_TYPE_Q4_0: 4   + QK/2,
+    GGML_TYPE_Q4_1: 4*2 + QK/2,
```
Great catch. Fixed.

jart commented on convert-pth-to-ggml.py:
```diff
@@ -95,85 +136,141 @@ def write_tokens(fout, tokenizer):
         print(f"Extracting only the vocab from '{fname_model}'\n")
-
-
+        model = torch.load(fname_model, map_location="cpu")
```
Thanks. Fixed!

x02Sylvie commented:
I don't really know python, so I'd rather leave pull request to someone smarter than me,

I did however manage to get alpaca 13b model converted by manually setting n_parts to 1 in .py conversion script . I'm unsure if it's proper place to set n_parts though
```
def get_n_parts(dim):
    mappings = {4096: 1, 5120: 2, 6656: 4, 8192: 8}

    n_parts = mappings.get(dim)

    if n_parts is None:
        print(f"Invalid dim: {dim}")
        sys.exit(1)
    print(f"n_parts = {n_parts}\n")
    return n_parts
```
to
```
def get_n_parts(dim):
    mappings = {4096: 1, 5120: 2, 6656: 4, 8192: 8}

    n_parts = 1

    if n_parts is None:
        print(f"Invalid dim: {dim}")
        sys.exit(1)
    print(f"n_parts = {n_parts}\n")
    return n_parts
```

jart commented:
Yes, that code which essentially guesses `n_parts` based off the dimension sizes looks like a LLaMA kludge to me. @ggerganov would need to weigh in before we change it. I was simply cargo culting other parts of the codebase that do this.

rabidcopy commented:
So far the conversion script works with my current ggml models. Even converted the gpt4all model with `convert-gpt4all-to-ggml.py` and then converted that with `migrate-ggml-2023-03-30-pr613.py` and it works. Do have to manually set n_parts in the script for my larger models that are in 1 part. But still works nonetheless!

jart commented:
I've just pushed an update to this PR addressing the `n_part` issue. We're now using this algorithm in the migration tool:

```py
    # count number of multipart files by convention
    n_parts = 1
    while True:
        if os.path.exists("%s.%d" % (args.fin_path, n_parts)):
            n_parts += 1
        else:
            break
```

ggerganov commented on llama.cpp:
```diff
@@ -288,6 +307,58 @@ struct llama_context_params llama_context_default_params() {
+            "\tsee https://github.com/ggerganov/llama.cpp/issues/91\n"
+            "\tuse convert-pth-to-ggml.py to regenerate from original pth\n"
+            "\tuse migrate-ggml-2023-03-30-pr613.py if you deleted originals\n",
+            path);
```
Let's print the magic we got and the magic we expect here to help debug issues

jart replied:
As you wish. Updated.

ggerganov reviewed (approved):
I think this is ready to merge - please go ahead

🦙

CoderRC commented:
Successfully compiled master branch and successfully compiled jart's master branch, and successfully run ./main -m ./models/7B/ggml-model-q4_0.bin -p "Building a website can be done in 10 simple steps:" -t 8 -n 512.
In msys2 with mingw32 gcc compiler using:
make LDFLAGS='-D_POSIX_MAPPED_FILES -lmingw32_extended' CFLAGS='-D_POSIX_MAPPED_FILES   -I.              -O3 -DNDEBUG -std=c11   -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -Wno-unused-function -mfma -mf16c -mavx -mavx2' CXXFLAGS='-D_POSIX_MAPPED_FILES -I. -I./examples -O3 -DNDEBUG -std=c++11 -fPIC -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function'

If confused how exactly I did compile it, read https://github.com/ggerganov/llama.cpp/issues/103#issuecomment-1487784446

CoderRC commented:
This should be ready to merge due to my testing and compiling not failing.

gaceladri commented:
Hello,

I can not load the gtp4all after converting it to the new ggml format using your script:
```python3 convert-gpt4all-to-ggml.py models/gpt4all/gpt4all-lora-quantized.bin ./models/tokenizer.model```

I have opened a new issue probably related to this: https://github.com/ggerganov/llama.cpp/issues/655#issue-1649732519

gaceladri commented:
I could run it with the previous version https://github.com/ggerganov/llama.cpp/tree/master-ed3c680 

> Hello,
> 
> I can not load the gtp4all after converting it to the new ggml format using your script: `python3 convert-gpt4all-to-ggml.py models/gpt4all/gpt4all-lora-quantized.bin ./models/tokenizer.model`
> 
> I have opened a new issue probably related to this: [#655 (comment)](https://github.com/ggerganov/llama.cpp/issues/655#issue-1649732519)

rabidcopy commented:
> Hello,
> 
> I can not load the gtp4all after converting it to the new ggml format using your script: `python3 convert-gpt4all-to-ggml.py models/gpt4all/gpt4all-lora-quantized.bin ./models/tokenizer.model`
> 
> I have opened a new issue probably related to this: [#655 (comment)](https://github.com/ggerganov/llama.cpp/issues/655#issue-1649732519)

You need to also run the resulting file through `migrate-ggml-2023-03-30-pr613.py` as well.

gaceladri commented:
It worked. Thank you for your fast response!

asklar commented:
great work @jart and @slaren ! <3
